pkg/kubelet/images/pullmanager/image_pull_manager.go

```diff
@@ -0,0 +1,479 @@
+	recordsAccessor PullRecordsAccessor
+
+	requireCredentialVerification ImagePullPolicyEnforcer
+
+	imageService kubecontainer.ImageService
+
+	intentAccessors *namedLockSet  // image -> sync.Mutex
+	intentCounters  map[string]int // image -> number of current in-flight pulls
+
+	pulledAccessors *namedLockSet // imageRef -> sync.Mutex
+}
+
+func NewImagePullManager(ctx context.Context, recordsAccessor PullRecordsAccessor, imagePullPolicy ImagePullPolicyEnforcer, imageService kubecontainer.ImageService) (*PullManager, error) {
+	m := &PullManager{
+		recordsAccessor: recordsAccessor,
+
+		requireCredentialVerification: imagePullPolicy,
+
+		imageService: imageService,
+
+		intentAccessors: NewNamedLockSet(),
+		intentCounters:  make(map[string]int),
+
+		pulledAccessors: NewNamedLockSet(),
+	}
+
+	m.initialize(ctx)
+
+	return m, nil
+}
+
+func (f *PullManager) RecordPullIntent(image string) error {
+	f.intentAccessors.Lock(image)
+	defer f.intentAccessors.Unlock(image)
+
+	if err := f.recordsAccessor.WriteImagePullIntent(image); err != nil {
+		return fmt.Errorf("failed to record image pull intent: %w", err)
+	}
+
+	f.intentCounters[image]++
+	return nil
+}
+
+func (f *PullManager) RecordImagePulled(image, imageRef string, credentials *kubeletconfiginternal.ImagePullCredentials) {
+	if credentials == nil || (len(credentials.KubernetesSecrets) == 0 && !credentials.NodePodsAccessible) {
+		klog.ErrorS(
+			fmt.Errorf("credentials must not be nil, and they must contain at least one kubernetes secret coordinate or must set NodePodsAccessible=true"),
+			"attempted to write an ImagePulledRecord with invalid credentials",
+			"image", image, "imageRef", imageRef, "credentials", credentials)
+		return
+	}
+
+	if err := f.writePulledRecord(image, imageRef, credentials); err != nil {
+		klog.ErrorS(err, "failed to write image pulled record", "imageRef", imageRef)
+		return
+	}
+
+	f.decrementImagePullIntent(image)
+}
+
+// writePulledRecord writes an ImagePulledRecord into the f.pulledDir directory.
+// `image` is an image from a container of a Pod object.
+// `imageRef` is a reference to the `image“ as used by the CRI.
+// `credentials` is a set of credentials that should be written to a new/merged into
+// an existing record.
+//
+// If `credentials` is nil, it marks a situation where an image was pulled under
+// unknown circumstances. We should record the image as tracked but no credentials
+// should be written in order to force credential verification when the image is
+// accessed the next time.
+func (f *PullManager) writePulledRecord(image, imageRef string, credentials *kubeletconfiginternal.ImagePullCredentials) error {
+	f.pulledAccessors.Lock(imageRef)
+	defer f.pulledAccessors.Unlock(imageRef)
+
+	sanitizedImage, err := trimImageTagDigest(image)
+	if err != nil {
+		return fmt.Errorf("invalid image name %q: %w", image, err)
+	}
+
+	pulledRecord, _, err := f.recordsAccessor.GetImagePulledRecord(imageRef)
+	if err != nil {
+		return err
```

Comment:
This is an interesting case... if we can't read the existing record for some reason, _should_ we error and return early? That would mean we'd fail to write the new record, and that a corrupted pulled record file cannot self heal, and will keep us from remembering new credentials used to pull it are valid

Reply:
This is indeed likely the best place to attempt self-heal. There may not be much value keeping a record we cannot automatically read.